### PR TITLE
Add include_only fields

### LINKED ISSUE #228: Add `include_only` and `include_only_regexp`
From 17405ea:

The CKAN filters (does not install) any file which _matches_ the regexp. The negative look-arounds means we can have a regexp which matches everything _except_ for a given file or path. Yes, that's needlessly complex.

To be honest, in addition to the filter fields, we should probably have their inverse as well, which I propose as `include_only` and `include_only_regexp`. It would be an error to have both filters and include_onlys in the same stanza.

That means we can have `"filter" : "config.cfg"` in the main document, and `"install_only" : "config.cfg"` in the configuration documents, which makes things much easier and more obviously correct.

### PR DESCRIPTION
**Pull request by Olympic1**
This PR adds two fields which practically work as negative filter/filter_regexp.
Also includes a test and changes to the schema to make sure there are only filter or include_only fields but never both.
Also added the new "filters" to the spec.

Closes #228
Closes #1577

### CODE REVIEW

File: Spec.md [hunk 1]
```diff
@@ -295,14 +295,22 @@ In addition, any number of optional directives *may* be provided:
...
-  expression, it is not installed.
+  expression, it is *not* installed.
+- `include_only` : (**v1.24**) A string, or list of strings, of file parts that should
+  be installed. These are treated as literal things which must match a
+  file name or directory. Examples of this may be `Settings.cfg`,
+  or `Plugin`. These are considered case-insensitive.
+- `include_only_regexp` : (**v1.24**) A string, or list of strings, which are treated as
+  case-sensitive C# regular expressions which are matched against the
+  full paths from the installing zip-file. If a file matches the regular
+  expression, it is installed.
```
Comment (politas) [inline Spec.md]: Maybe bold the "is" to contrast the "not" above?

**Comment by politas** [comment]
The logic seems good to me. I'll do some testing

File: Core/Types/ModuleInstallDescriptor.cs [hunk 2]
```diff
@@ -65,6 +73,15 @@ internal void DeSerialisationFixes(StreamingContext like_i_could_care)
...
             }
+
+            // Make sure only filter or include_only fields exist but not both at the same time
+            var filterCount = new[] { filter, filter_regexp }.Count(i => i != null);
+            var includeOnlyCount = new[] { include_only, include_only_regexp }.Count(i => i != null);
+
+            if (filterCount > 0 && includeOnlyCount > 0)
+            {
+                throw new BadMetadataKraken(null, "Install stanzas can only contain filter or include_only directives, not both");
+            }
```
Comment (HebaruSan) [inline Core/Types/ModuleInstallDescriptor.cs]: Could we allow both? I could see that being useful.

```
"include_only": "Astrogator.*"
"filter": "*.md"
```

Comment (politas) [reply Core/Types/ModuleInstallDescriptor.cs]: Which would take preference? When both a filter and an include_only match, there's no  clear decision. We don't have operator ordering in the spec.

Comment (HebaruSan) [reply Core/Types/ModuleInstallDescriptor.cs]: I was thinking that if both match, that would be treated as an exclude.

Comment (politas) [reply Core/Types/ModuleInstallDescriptor.cs]: I think it is better to use separate install clauses to cover such cases. I think that gives us sufficient flexibility, without the confusing interactions of competing filters.

**Comment by politas** [comment]
Ok, well that seems to work pretty well - putting the following install elements in DCVDavincisCombatKit.netkan worked exactly as I wanted it to:

    "install": [
        {
            "find_regexp": "GameData",
            "install_to": "GameData/DCK",
            "include_only": [ "DCK.version", "GPLV3.txt", "LICENSE", "License.txt", "README.md" ]
        },
        {
            "find_regexp": "GameData",
            "install_to": "GameData",
            "include_only_regexp": "DCK_.*"
        }
    ]

**Comment by politas** [comment]
Great work, @Olympic1!

**Comment by HebaruSan** [comment]
A pretty mundane commit in https://github.com/KSP-CKAN/NetKAN/pull/6015 is erroring out in the schema validation step now. Is there a syntax error in the new changes? Or maybe the python script needs an update?

```
Checking built/ActionGroupManager-2.2.1.ckan
Validating built/ActionGroupManager-2.2.1.ckan..
Traceback (most recent call last):
  File "./ckan-validate.py", line 58, in <module>
    main()
  File "./ckan-validate.py", line 42, in main
    validate(json.load(ckan_file), schema)
  File "/usr/lib/python2.7/dist-packages/jsonschema/validators.py", line 431, in validate
    cls.check_schema(schema)
  File "/usr/lib/python2.7/dist-packages/jsonschema/validators.py", line 76, in check_schema
    raise SchemaError.create_from(error)
jsonschema.exceptions.SchemaError: {u'not': {u'anyOf': [{u'required': [u'filter', u'include_only']}, {u'required': [u'filter', u'include_only_regexp']}, {u'required': [u'filter_regexp', u'include_only']}, {u'required': [u'filter_regexp', u'include_only_regexp']}]}, u'oneOf': [{u'required': [u'file']}, {u'required': [u'find']}, {u'required': [u'find_regexp']}], u'required': [u'install_to'], u'type': u'object', u'properties': {u'find_regexp': {u'type': u'string', u'description': u'A regexp that matches the directory to install'}, u'find_matches_files': {u'type': u'boolean', u'description': u'If true, find directives match files as well as directories'}, u'filter_regexp': {u'description': u'List of regexps that should filter files from this install', u'$ref': u'#/definitions/one_or_more_strings'}, u'filter': {u'description': u'List of files and directories that should be filtered from the install', u'$ref': u'#/definitions/one_or_more_strings'}, u'as': {u'pattern': u'^[^\\\\/]+$', u'type': u'string', u'description': u'The name to give the matching directory or file when installed'}, u'install_to': {u'description': u'Where file should be installed to', u'$ref': u'#/definitions/install_to'}, u'file': {u'type': u'string', u'description': u'Path to file to install'}, u'include_only_regexp': {u'type': u'#/definitions/one_or_more_strings', u'description': u'List of regexps that should include files in this install'}, u'find': {u'type': u'string', u'description': u'A directory to find for installation'}, u'include_only': {u'type': u'#/definitions/one_or_more_strings', u'description': u'List of files and directories that should not be excluded from the install'}}} is not valid under any of the given schemas

Failed validating u'anyOf' in schema[u'properties'][u'properties'][u'additionalProperties'][u'properties'][u'items']:
    {u'anyOf': [{u'$ref': u'#'}, {u'$ref': u'#/definitions/schemaArray'}],
     u'default': {}}

On instance[u'properties'][u'install'][u'items']:
    {u'not': {u'anyOf': [{u'required': [u'filter', u'include_only']},
                         {u'required': [u'filter',
                                        u'include_only_regexp']},
                         {u'required': [u'filter_regexp',
                                        u'include_only']},
                         {u'required': [u'filter_regexp',
                                        u'include_only_regexp']}]},
     u'oneOf': [{u'required': [u'file']},
                {u'required': [u'find']},
                {u'required': [u'find_regexp']}],
     u'properties': {u'as': {u'description': u'The name to give the matching directory or file when installed',
                             u'pattern': u'^[^\\\\/]+$',
                             u'type': u'string'},
                     u'file': {u'description': u'Path to file to install',
                               u'type': u'string'},
                     u'filter': {u'$ref': u'#/definitions/one_or_more_strings',
                                 u'description': u'List of files and directories that should be filtered from the install'},
                     u'filter_regexp': {u'$ref': u'#/definitions/one_or_more_strings',
                                        u'description': u'List of regexps that should filter files from this install'},
                     u'find': {u'description': u'A directory to find for installation',
                               u'type': u'string'},
                     u'find_matches_files': {u'description': u'If true, find directives match files as well as directories',
                                             u'type': u'boolean'},
                     u'find_regexp': {u'description': u'A regexp that matches the directory to install',
                                      u'type': u'string'},
                     u'include_only': {u'description': u'List of files and directories that should not be excluded from the install',
                                       u'type': u'#/definitions/one_or_more_strings'},
                     u'include_only_regexp': {u'description': u'List of regexps that should include files in this install',
                                              u'type': u'#/definitions/one_or_more_strings'},
                     u'install_to': {u'$ref': u'#/definitions/install_to',
                                     u'description': u'Where file should be installed to'}},
     u'required': [u'install_to'],
     u'type': u'object'}
Build step 'Execute shell' marked build as failure
```

**Comment by HebaruSan** [comment]
Shortening to highlight the error:
```
jsonschema.exceptions.SchemaError: {u'not': {u'anyOf': [{u'required': [u'filter', u'include_only']}, {u'required': [u'filter', u'include_only_regexp']}, ...}
is not valid under any of the given schemas
```
It doesn't seem to like that `"not": {"anyOf"...` block.

**Pull request merged**